// @usebruno/js/src/bru.js
const { cloneDeep } = require('lodash');
const { uuid } = require('./utils');
const xmlFormat = require('xml-formatter');
const { interpolate: _interpolate } = require('@usebruno/common');
const { createSendRequest } = require('@usebruno/requests').scripting;
const { jar: createCookieJar, getCookiesForUrl } = require('@usebruno/requests').cookies;
const CookieList = require('./cookie-list');
const Handlebars = require('handlebars');

const VISUALIZATION_CLEAR = '__clear__';
const MAX_TEMPLATE_LENGTH = 1_000_000; // 1 million characters
/*
 * All Handlebars.compile() options (v4.x):
 *   data                    - set false to disable @data tracking
 *   compat                  - enable recursive field lookup
 *   knownHelpers            - hash of helpers known at compile time (optimization)
 *   knownHelpersOnly        - only allow known helpers (optimization)       [SAFE]
 *   noEscape                - skip HTML escaping of content                 [SAFE]
 *   strict                  - throw on missing fields instead of ignoring   [SAFE]
 *   assumeObjects           - skip object existence checks on paths         [SAFE]
 *   preventIndent           - disable auto-indent for partials              [SAFE]
 *   ignoreStandalone        - disable standalone tag removal                [SAFE]
 *   explicitPartialContext  - disable implicit context for partials         [SAFE]
 *
 * Excluded from allowlist:
 *   data, compat, knownHelpers - can alter template resolution in unexpected ways
 *   allowProtoPropertiesByDefault, allowProtoMethodsByDefault - defense-in-depth against
 *     prototype traversal. Handlebars >=4.3.0 fixed the original CVE-2019-19919 and >=4.7.0
 *     hardcodes blocks on __proto__, constructor, etc. even when these flags are true, but
 *     we still exclude them to avoid relaxing the security posture.
 *     https://nvd.nist.gov/vuln/detail/CVE-2019-19919
 *
 * Use `dangerouslyAllowAllOptions: true` in data.options to bypass the allowlist.
 */
const SAFE_HBS_OPTIONS = new Set([
  'noEscape', 'strict', 'assumeObjects', 'preventIndent',
  'ignoreStandalone', 'explicitPartialContext', 'knownHelpersOnly'
]);
const variableNameRegex = /^[\w-.]*$/;

class Bru {
  /**
   * @param {object} options - Single options object (destructured)
   * @property {string} options.runtime - The runtime environment ('quickjs' or 'nodevm')
   * @property {object} [options.envVariables={}] - Environment variables
   * @property {object} [options.runtimeVariables={}] - Runtime variables
   * @property {object} [options.processEnvVars={}] - Process environment variables (deep cloned)
   * @property {string} [options.collectionPath] - Path to the collection
   * @property {function} [options.historyLogger] - History logger function
   * @property {function} [options.setVisualizations] - Visualizations setter function
   * @property {object} [options.secretVariables={}] - Secret variables (deep cloned)
   * @property {object} [options.collectionVariables={}] - Collection-level variables
   * @property {object} [options.folderVariables={}] - Folder-level variables
   * @property {object} [options.requestVariables={}] - Request-level variables
   * @property {object} [options.globalEnvironmentVariables={}] - Global environment variables
   * @property {object} [options.oauth2CredentialVariables={}] - OAuth2 credential variables
   * @property {object} [options.iterationDetails] - Iteration details for runner
   * @property {string} [options.collectionName] - Name of the collection
   * @property {object} [options.promptVariables={}] - Prompt variables
   * @property {object} [options.certsAndProxyConfig] - Configuration for bru.sendRequest (proxy, certs, TLS)
   * @property {string} [options.certsAndProxyConfig.collectionPath] - Path to the collection
   * @property {object} [options.certsAndProxyConfig.options] - TLS and proxy options
   * @property {object} [options.certsAndProxyConfig.clientCertificates] - Client certificate configuration
   * @property {object} [options.certsAndProxyConfig.collectionLevelProxy] - Collection-level proxy settings
   * @property {object} [options.certsAndProxyConfig.systemProxyConfig] - System proxy configuration
   * @property {string} [options.requestUrl] - The URL of the current request (used for cookie access)
   * @property {function} [options.onConsoleLog] - Console callback `(type, args) => void` to route warnings to the UI
   */
  constructor({
    runtime,
    envVariables,
    runtimeVariables,
    processEnvVars,
    collectionPath,
    historyLogger,
    setVisualizations,
    secretVariables,
    collectionVariables,
    folderVariables,
    requestVariables,
    globalEnvironmentVariables,
    oauth2CredentialVariables,
    iterationDetails,
    collectionName,
    promptVariables,
    certsAndProxyConfig,
    requestUrl,
    onConsoleLog
  }) {
    this.envVariables = envVariables || {};
    this.runtimeVariables = runtimeVariables || {};
    this.promptVariables = promptVariables || {};
    this.processEnvVars = cloneDeep(processEnvVars || {});
    this.secretVariables = cloneDeep(secretVariables || {});
    this.collectionVariables = collectionVariables || {};
    this.folderVariables = folderVariables || {};
    this.requestVariables = requestVariables || {};
    this.globalEnvironmentVariables = globalEnvironmentVariables || {};
    this.oauth2CredentialVariables = oauth2CredentialVariables || {};
    this.collectionPath = collectionPath;
    this.historyLogger = historyLogger;
    this.setVisualizations = setVisualizations;
    this.onConsoleLog = onConsoleLog || null;
    this.collectionName = collectionName;
    // Set by the host-side __bruSetScope global at the top of each segment's IIFE.
    this._currentScope = null;
    this.scriptedRequestEntries = [];
    this.sendRequest = (...args) => {
      const scopeSnapshot = this._currentScope ? { ...this._currentScope } : null;
      const send = createSendRequest(certsAndProxyConfig, {
        onComplete: (entry) =>
          this._recordScriptedRequest({ source: 'sendRequest', scope: scopeSnapshot, ...entry })
      });
      return send(...args);
    };
    this.runtime = runtime;
    this.requestUrl = requestUrl;
    this.cookies = new CookieList({
      getUrl: () => this.interpolate(this.requestUrl),
      interpolate: (str) => this.interpolate(str),
      createCookieJar,
      getCookiesForUrl
    });
    // Holds variables that are marked as persistent by scripts
    this.persistentEnvVariables = {};
    // Holds credential IDs to be reset after script execution
    this.oauth2CredentialsToReset = [];
    this.runner = {
      skipRequest: () => {
        this.skipRequest = true;
      },
      stopExecution: () => {
        this.stopExecution = true;
      },
      setNextRequest: (nextRequest) => {
        this.nextRequest = nextRequest;
      },
      iterationData: new IterationDataManager(iterationDetails?.iterationData),
      iterationIndex: iterationDetails?.iterationIndex,
      totalIterations: iterationDetails?.totalIterations
    };

    this.utils = {
      minifyJson: (json) => {
        if (json === null || json === undefined) {
          throw new Error('Failed to minify');
        }

        if (typeof json === 'object') {
          try {
            return JSON.stringify(json);
          } catch (err) {
            throw new Error(`Failed to minify: ${err?.message || err}`);
          }
        }

        if (typeof json === 'string') {
          const trimmed = json.trim();
          if (trimmed === '') return trimmed;
          try {
            return JSON.stringify(JSON.parse(trimmed));
          } catch (err) {
            throw new Error(`Failed to minify: ${err?.message || err}`);
          }
        }

        throw new TypeError('minifyJson expects a string or object');
      },

      minifyXml: (xml) => {
        if (xml === null || xml === undefined) {
          throw new Error('Failed to minify');
        }

        if (typeof xml === 'string') {
          try {
            return xmlFormat(xml, { collapseContent: false, indentation: '', lineSeparator: '' });
          } catch (err) {
            throw new Error(`Failed to minify: ${err?.message || err}`);
          }
        }

        throw new TypeError('minifyXml expects a string');
      }
    };
  }

  interpolate = (strOrObj) => {
    if (!strOrObj) return strOrObj;
    const isObj = typeof strOrObj === 'object';
    const strToInterpolate = isObj ? JSON.stringify(strOrObj) : strOrObj;

    const combinedVars = {
      ...this.globalEnvironmentVariables,
      ...this.collectionVariables,
      ...this.envVariables,
      ...this.folderVariables,
      ...this.requestVariables,
      ...this.oauth2CredentialVariables,
      ...this.runtimeVariables,
      ...this.promptVariables,
      process: {
        env: {
          ...this.processEnvVars
        }
      }
    };

    const interpolatedStr = _interpolate(strToInterpolate, combinedVars);
    return isObj ? JSON.parse(interpolatedStr) : interpolatedStr;
  };

  cwd() {
    return this.collectionPath;
  }

  _recordScriptedRequest(entry) {
    // Prefer scope passed in by the caller (snapshot at call time). Fall back to
    // _currentScope for callers that don't supply one (e.g. bru.runRequest).
    const { scope: providedScope, ...rest } = entry;
    const scope = providedScope !== undefined
      ? providedScope
      : (this._currentScope ? { ...this._currentScope } : null);
    this.scriptedRequestEntries.push({ ...rest, scope });
  }

  getEnvName() {
    return this.envVariables.__name__;
  }

  getProcessEnv(key) {
    return this.processEnvVars[key];
  }

  hasEnvVar(key) {
    return Object.hasOwn(this.envVariables, key);
  }

  getEnvVar(key) {
    return this.interpolate(this.envVariables[key]);
  }

  setEnvVar(key, value, options = {}) {
    if (!key) {
      throw new Error('Creating a env variable without specifying a name is not allowed.');
    }

    if (variableNameRegex.test(key) === false) {
      throw new Error(
        `Variable name: "${key}" contains invalid characters! Names must only contain alpha-numeric characters, "-", "_", "."`
      );
    }

    // When persist is true, only string values are allowed
    if (options?.persist && typeof value !== 'string') {
      throw new Error(`Persistent environment variables must be strings. Received ${typeof value} for key "${key}".`);
    }

    this.envVariables[key] = value;

    if (options?.persist) {
      this.persistentEnvVariables[key] = value;
    } else {
      if (this.persistentEnvVariables[key]) {
        delete this.persistentEnvVariables[key];
      }
    }
  }

  deleteEnvVar(key) {
    delete this.envVariables[key];
  }

  getAllEnvVars() {
    const vars = Object.assign({}, this.envVariables);
    delete vars.__name__;
    return vars;
  }

  deleteAllEnvVars() {
    const envName = this.envVariables.__name__;
    for (let key in this.envVariables) {
      if (this.envVariables.hasOwnProperty(key)) {
        delete this.envVariables[key];
      }
    }
    if (envName !== undefined) {
      this.envVariables.__name__ = envName;
    }
  }

  hasGlobalEnvVar(key) {
    return Object.hasOwn(this.globalEnvironmentVariables, key);
  }

  getGlobalEnvVar(key) {
    return this.interpolate(this.globalEnvironmentVariables[key]);
  }

  setGlobalEnvVar(key, value) {
    if (!key) {
      throw new Error('Creating a env variable without specifying a name is not allowed.');
    }

    this.globalEnvironmentVariables[key] = value;
  }

  // TODO: deleteGlobalEnvVar works in the request lifecycle but does not update the UI.
  // Re-enable once the UI sync issue is resolved.
  // deleteGlobalEnvVar(key) {
  //   delete this.globalEnvironmentVariables[key];
  // }

  getAllGlobalEnvVars() {
    return Object.assign({}, this.globalEnvironmentVariables);
  }

  // TODO: deleteAllGlobalEnvVars works in the request lifecycle but does not update the UI.
  // Re-enable once the UI sync issue is resolved.
  // deleteAllGlobalEnvVars() {
  //   for (let key in this.globalEnvironmentVariables) {
  //     if (this.globalEnvironmentVariables.hasOwnProperty(key)) {
  //       delete this.globalEnvironmentVariables[key];
  //     }
  //   }
  // }

  getOauth2CredentialVar(key) {
    return this.interpolate(this.oauth2CredentialVariables[key]);
  }

  resetOauth2Credential(credentialId) {
    if (!credentialId || typeof credentialId !== 'string') {
      throw new Error('credentialId must be a non-empty string');
    }

    if (!this.oauth2CredentialsToReset.includes(credentialId)) {
      this.oauth2CredentialsToReset.push(credentialId);
    }

    // Remove matching credential variables so subsequent getOauth2CredentialVar() calls return undefined
    const prefix = `$oauth2.${credentialId}.`;
    for (const key of Object.keys(this.oauth2CredentialVariables)) {
      if (key.startsWith(prefix)) {
        delete this.oauth2CredentialVariables[key];
      }
    }
  }

  hasVar(key) {
    return Object.hasOwn(this.runtimeVariables, key);
  }

  setVar(key, value) {
    if (!key) {
      throw new Error('Creating a variable without specifying a name is not allowed.');
    }

    if (variableNameRegex.test(key) === false) {
      throw new Error(
        `Variable name: "${key}" contains invalid characters!`
        + ' Names must only contain alpha-numeric characters, "-", "_", "."'
      );
    }

    this.runtimeVariables[key] = value;
  }

  getVar(key) {
    if (variableNameRegex.test(key) === false) {
      throw new Error(
        `Variable name: "${key}" contains invalid characters!`
        + ' Names must only contain alpha-numeric characters, "-", "_", "."'
      );
    }

    return this.interpolate(this.runtimeVariables[key]);
  }

  deleteVar(key) {
    delete this.runtimeVariables[key];
  }

  deleteAllVars() {
    for (let key in this.runtimeVariables) {
      if (this.runtimeVariables.hasOwnProperty(key)) {
        delete this.runtimeVariables[key];
      }
    }
  }

  getAllVars() {
    return Object.assign({}, this.runtimeVariables);
  }

  getCollectionVar(key) {
    return this.interpolate(this.collectionVariables[key]);
  }

  // TODO: setCollectionVar works in the request lifecycle but does not update the UI.
  // Re-enable once the UI sync issue is resolved.
  // setCollectionVar(key, value) {
  //   if (!key) {
  //     throw new Error('Creating a variable without specifying a name is not allowed.');
  //   }
  //
  //   if (variableNameRegex.test(key) === false) {
  //     throw new Error(
  //       `Variable name: "${key}" contains invalid characters!`
  //       + ' Names must only contain alpha-numeric characters, "-", "_", "."'
  //     );
  //   }
  //
  //   this.collectionVariables[key] = value;
  // }

  hasCollectionVar(key) {
    return Object.hasOwn(this.collectionVariables, key);
  }

  // TODO: deleteCollectionVar works in the request lifecycle but does not update the UI.
  // Re-enable once the UI sync issue is resolved.
  // deleteCollectionVar(key) {
  //   delete this.collectionVariables[key];
  // }

  // TODO: deleteAllCollectionVars works in the request lifecycle but does not update the UI.
  // Re-enable once the UI sync issue is resolved.
  // deleteAllCollectionVars() {
  //   for (let key in this.collectionVariables) {
  //     if (this.collectionVariables.hasOwnProperty(key)) {
  //       delete this.collectionVariables[key];
  //     }
  //   }
  // }

  // TODO: getAllCollectionVars works in the request lifecycle but does not update the UI.
  // Re-enable once the UI sync issue is resolved.
  // getAllCollectionVars() {
  //   return Object.assign({}, this.collectionVariables);
  // }

  getFolderVar(key) {
    return this.interpolate(this.folderVariables[key]);
  }

  getRequestVar(key) {
    return this.interpolate(this.requestVariables[key]);
  }

  setNextRequest(nextRequest) {
    this.nextRequest = nextRequest;
  }

  getSecretVar(key) {
    return this.secretVariables?.[`$secrets.${key}`];
  }

  /**
   * Render a visualization in the response pane.
   *
   * @param {'table' | 'html'} type - The visualization type.
   * @param {Object} data - Shape depends on `type`:
   *
   *   **When `type` is `'table'`:**
   *   @param {string}  data.provider             - `'ag-grid'` or `'react-table'`.
   *   @param {Object}  data.props                - Props forwarded to the grid component.
   *   @param {Array}   data.props.columnDefinitions - (ag-grid) Column definitions.
   *   @param {Array}   data.props.rowData           - (ag-grid) Row data.
   *
   *   **When `type` is `'html'`** (raw mode):
   *   @param {string}  data.content              - Pre-rendered HTML string.
   *
   *   **When `type` is `'html'`** (Handlebars template mode):
   *   @param {string}  data.template             - Handlebars template string (required).
   *   @param {Object}  [data.data={}]            - Context object passed to the compiled template.
   *   @param {Object}  [data.options]            - Handlebars compile options.
   *   @param {boolean} [data.options.noEscape]    - Disable HTML escaping.
   *   @param {boolean} [data.options.strict]      - Throw on missing variables.
   *   @param {boolean} [data.options.assumeObjects] - Skip null checks on nested paths.
   *   @param {boolean} [data.options.dangerouslyAllowAllOptions] - Bypass the safe-options
   *     allowlist (forwards all remaining options directly to `Handlebars.compile`).
   *
   * @throws {Error} If `type` is falsy.
   * @throws {Error} If `type` is `'html'` and neither `data.content` nor `data.template` is provided.
   * @throws {Error} If template exceeds `MAX_TEMPLATE_LENGTH`.
   */
  visualize(type, data) {
    if (!type) {
      throw new Error('visualization type is required');
    }
    if (type === 'table') {
      if (data?.provider === 'ag-grid') {
        if (!data?.props?.columnDefinitions) {
          throw new Error(`columns definitions are required`);
        }
        if (!data?.props?.rowData) {
          throw new Error(`row data is required`);
        }
        this.setVisualizations({ uid: uuid(), type, data });
      } else if (data?.provider === 'react-table') {
        this.setVisualizations({ uid: uuid(), type, data });
      }
    } else if (type === 'html') {
      if (typeof data?.template === 'string') {
        // Handlebars template mode (translated from pm.visualizer.set)
        if (data.template.length > MAX_TEMPLATE_LENGTH) {
          throw new Error(`template exceeds maximum length of ${MAX_TEMPLATE_LENGTH} characters`);
        }
        // Isolated instance so globally-registered helpers can't be exploited (SSTI prevention)
        const hbs = Handlebars.create();
        // Allowlist safe compile options to prevent re-enabling prototype traversal (CVE-2019-19919)
        let compileOptions;
        if (data.options?.dangerouslyAllowAllOptions) {
          const { dangerouslyAllowAllOptions, ...rest } = data.options;
          compileOptions = rest;
        } else {
          compileOptions = {};
          if (data.options) {
            const blocked = [];
            for (const key of Object.keys(data.options)) {
              if (SAFE_HBS_OPTIONS.has(key)) {
                compileOptions[key] = data.options[key];
              } else {
                blocked.push(key);
              }
            }
            if (blocked.length > 0) {
              const msg = `bru.visualize: the following compile options were blocked and ignored: ${blocked.map((k) => `"${k}"`).join(', ')}. `
                + 'Use { dangerouslyAllowAllOptions: true } in options to bypass the allowlist.';
              if (this.onConsoleLog) {
                this.onConsoleLog('warn', [msg]);
              } else {
                console.warn(msg);
              }
            }
          }
        }
        const compiled = hbs.compile(data.template, compileOptions);
        const renderedHtml = compiled(data.data || {});
        this.setVisualizations({
          uid: uuid(), type,
          data: { content: renderedHtml, rawData: data.data || {} }
        });
      } else if (data?.content) {
        // Existing raw HTML mode
        this.setVisualizations({ uid: uuid(), type, data });
      } else {
        throw new Error('html content or template is required');
      }
    }
  }

  clearVisualizations() {
    this.setVisualizations({ type: VISUALIZATION_CLEAR });
  }

  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  getCollectionName() {
    return this.collectionName;
  }

  isSafeMode() {
    return this.runtime === 'quickjs';
  }
}

class IterationDataManager {
  constructor(iterationData) {
    this._data = iterationData;

    if (!this._data) {
      this._data = {};
    }
  }

  set(key, value) {
    if (!this._data) {
      this._data = {};
    }

    this._data[key] = value;
  }

  get(key) {
    if (!key) {
      return this._data || {};
    }
    return this._data?.[key];
  }

  has(key) {
    return this._data && key in this._data;
  }

  stringify() {
    return JSON.stringify(this.get());
  }

  unset(key) {
    if (this._data[key]) {
      delete this._data[key];
    }
  }
}

module.exports = Bru;
module.exports.VISUALIZATION_CLEAR = VISUALIZATION_CLEAR;
