// grpc-js-reflection-client/dist/GrpcReflection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcReflection = void 0;
const Exceptions_1 = require("./Exceptions");
const protobufjs = __importStar(require("protobufjs"));
const Descriptor_1 = require("./Descriptor");
const descriptor_1 = require("protobufjs/ext/descriptor");
const lodash_1 = require("lodash");
const v1 = __importStar(require("./Proto/v1"));
const v1alpha = __importStar(require("./Proto/v1alpha"));
class GrpcReflection {
    client;
    version;
    reflectionRequestConstructor;
    constructor(host, credentials, options = {}, version = "v1alpha") {
        this.version = version;
        this.setProtoReflectionClient(host, credentials, options);
    }
    async listServices(prefix = '*', metadata, options) {
        const response = await this.request({
            listServices: prefix
        }, metadata, options);
        return response.listServicesResponse.service.map(service => service.name);
    }
    async listMethods(service, metadata, options) {
        const descriptor = await this.getDescriptorBySymbol(service, metadata, options);
        const packageObject = descriptor.getPackageObject({
            keepCase: true,
            enums: String,
            longs: String
        });
        return this.getServiceMethods(packageObject, service);
    }
    generateServicePath(service) {
        return service.split('.').reverse();
    }
    getServiceMethods(descriptor, service) {
        let anti_recursive = 0;
        let actualDescriptor = descriptor;
        const path = this.generateServicePath(service);
        do {
            const serviceName = path.pop();
            if (serviceName && serviceName in actualDescriptor) {
                actualDescriptor = actualDescriptor[serviceName];
            }
            anti_recursive++;
        } while (path.length > 0 && anti_recursive < 100);
        if ('service' in actualDescriptor) {
            return Object.entries(actualDescriptor.service)
                .map(([methodName, methodDefinition]) => ({
                name: methodName,
                definition: methodDefinition
            }));
        }
        throw new Exceptions_1.ReflectionRequestException('Not found service');
    }
    async getDescriptorByFileName(file_name, metadata, options) {
        const descriptor = await this.getProtoDescriptorByFileName(file_name, metadata, options);
        return await this.resolveFileDescriptorSet(descriptor, metadata, options);
    }
    async getDescriptorBySymbol(symbol, metadata, options) {
        const descriptor = await this.getProtoDescriptorBySymbol(symbol, metadata, options);
        return await this.resolveFileDescriptorSet(descriptor, metadata, options);
    }
    async getAllExtensionNumbersOfType(package_type, metadata, options) {
        const response = await this.request({
            allExtensionNumbersOfType: package_type
        }, metadata, options);
        return {
            base_type_name: response.allExtensionNumbersResponse.baseTypeName,
            extension_number: typeof (response.allExtensionNumbersResponse.extensionNumber) !== "undefined" ? response.allExtensionNumbersResponse.extensionNumber : []
        };
    }
    async request(payload, metadata, options) {
        return new Promise((resolve, reject) => {
            const payloadObject = this.reflectionRequestConstructor(payload);
            const call = this.client.ServerReflectionInfo(metadata, options);
            call.on('data', (data) => {
                if (data.errorResponse) {
                    reject(new Exceptions_1.ReflectionRequestException(data.errorResponse.errorMessage));
                    return;
                }
                resolve(data);
            });
            call.on('error', (err) => {
                reject(new Exceptions_1.ReflectionRequestException(err));
            });
            call.on('end', () => { });
            call.write(payloadObject);
            call.end();
        });
    }
    async resolveFileDescriptorSet(fileDescriptorProtoBytes, metadata, options) {
        const fileDescriptorSet = descriptor_1.FileDescriptorSet.create();
        const fileDescriptorProtos = await this.resolveDescriptorRecursive(fileDescriptorProtoBytes, metadata, options);
        (0, lodash_1.set)(fileDescriptorSet, 'file', Array.from(fileDescriptorProtos.values()));
        return new Descriptor_1.Descriptor(protobufjs.Root.fromDescriptor(fileDescriptorSet));
    }
    async resolveDescriptorRecursive(fileDescriptorProtoBytes, metadata, options) {
        let fileDescriptorProtos = new Map();
        let needsDependencyResolution = new Set();
        for (const item of fileDescriptorProtoBytes) {
            const fileDescriptorProto = descriptor_1.FileDescriptorProto.decode(item);
            if (fileDescriptorProto.dependency) {
                const dependencies = fileDescriptorProto.dependency;
                for (const dep of dependencies) {
                    needsDependencyResolution.add(dep);
                }
            }
            if (!fileDescriptorProtos.has(fileDescriptorProto.name)) {
                fileDescriptorProtos.set(fileDescriptorProto.name, fileDescriptorProto);
            }
        }
        for (const dep of needsDependencyResolution) {
            if (fileDescriptorProtos.has(dep)) {
                continue;
            }
            const depProtoBytes = await this.getProtoDescriptorByFileName(dep, metadata, options);
            const protoDependencies = await this.resolveDescriptorRecursive(depProtoBytes, metadata, options);
            fileDescriptorProtos = new Map([
                ...fileDescriptorProtos,
                ...protoDependencies,
            ]);
        }
        return fileDescriptorProtos;
    }
    async getProtoDescriptorBySymbol(symbol, metadata, options) {
        const response = await this.request({
            fileContainingSymbol: symbol
        }, metadata, options);
        return response.fileDescriptorResponse.fileDescriptorProto;
    }
    async getProtoDescriptorByFileName(file_name, metadata, options) {
        const response = await this.request({
            fileByFilename: file_name
        }, metadata, options);
        return response.fileDescriptorResponse.fileDescriptorProto;
    }
    setProtoReflectionClient(host, credentials, options = {}) {
        switch (this.version) {
            case 'v1':
                this.client = new v1.grpc.reflection.v1.ServerReflectionClient(host, credentials, options);
                this.reflectionRequestConstructor = v1.grpc.reflection.v1.ServerReflectionRequest.fromObject;
                break;
            case 'v1alpha':
                this.client = new v1alpha.grpc.reflection.v1alpha.ServerReflectionClient(host, credentials, options);
                this.reflectionRequestConstructor = v1alpha.grpc.reflection.v1alpha.ServerReflectionRequest.fromObject;
                break;
            default:
                throw new Exceptions_1.ReflectionRequestException('Unknown proto version available: [v1, v1alpha]');
        }
    }
}
exports.GrpcReflection = GrpcReflection;
